CRAYSAT-1459: Make --base* args optional, fix help text

jack-stanek-hpe opened the pull request:
## Summary and Scope

This change fixes the help text for the --base* options regarding their
behavior when no valid base is given. The text specifying that
cfs-config-util will start with an empty configuration has been deleted.
It is left implicit that non-existent configurations and files will
raise errors.

This change also modifies the behavior of cfs-config-util when no
--base* option is given. When no base is given, cfs-config-util will
start from an empty configuration with no layers. --save and
--save-suffix may not be used if no --base* is given. To avoid
clobbering existing configurations, --save-to-cfs and --save-to-file
will not work if an existing configuration or filename is given.


## Issues and Related PRs

* Resolves [CRAYSAT-1459](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1459)

## Testing

### Tested on:

  * `groot`

### Test description:

Run unit tests. On groot, manually do a sanity check by running
cfs-config-util commands inside container with various combinations of
options: without --base-* and with --save-* options, both once and
multiple times to check overwrite behavior; check with --base-* and with
--save-* to check for regressions; Check with non-existent --base-*
options.

## Risks and Mitigations

Creating empty configurations could overwrite existing configurations. Feature was designed to prevent data loss and is tested.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

haasken-hpe commented on cfs_config_util/cfs.py:
```diff
@@ -376,24 +387,42 @@ def req_payload(self):
+        if not overwrite:
+            try:
+                response = self._cfs_client.get('v2', 'configurations', name, raise_not_ok=False)
+                if response.status_code != 404:
+                    raise CFSConfigurationError(f'A configuration named {name} already exists '
+                                                f'and may not be overwritten.')
```
This condition will raise the `CFSConfigurationError` with a message that the configuration already exists when the `response.status_code` is not ok, but not 404. I think we really only want to raise an "Already exists" error when the response is ok and some other error when the response is not ok but not 404. Something like:

```suggestion
                # If the response is okay, it already exists
                if response.ok:
                    raise CFSConfigurationError(f'A configuration named {name} already exists '
                                                f'and may not be overwritten.')
                # Response of 404 Not Found is okay
                elif response.status_code != 404:
                    raise APIError(f'Failed to retrieve CFS configuration "{self.name}": '
                                   f'CFS returned status code {response.status_code}: {response.reason}')
```

haasken-hpe replied:
As an aside, there are a few places we should do something like this in the `sat` package as well. Specifically, I have some TODO comments about this in `sat bootprep`.

jack-stanek-hpe replied:
I agree, I believe there's an instance of this in `sat swap blade` too relating to waiting for NodeBMCs to be discovered. I can create a JIRA to port the `APIClient` API changes back to SAT and use that for those two instances.

haasken-hpe replied:
Thanks!

haasken-hpe commented on cfs_config_util/cfs.py:
```diff
@@ -376,24 +387,42 @@ def req_payload(self):
+
+            except APIError as err:
+                raise CFSConfigurationError(f'Failed to retrieve CFS configuration "{self.name}": {err}')
+            except ValueError as err:
+                raise CFSConfigurationError(f'Failed to decode JSON response from retrieving '
+                                            f'CFS configuration "{self.name}": {err}')
```
I don't believe this is needed as we're not calling `response.json` and don't need to. We just want to check the status code like you're doing.

haasken-hpe commented on cfs_config_util/cfs.py:
```diff
@@ -406,22 +435,28 @@ def save_to_cfs(self, name=None):
                 json.dump(self.req_payload, f, indent=2)
+        except FileExistsError:
+            raise CFSConfigurationError(f'Configuration at path {file_path} already exists '
+                                        f'and may not be overwritten.')
```
minor suggestion

```suggestion
                                        f'and will not be overwritten.')
```

haasken-hpe reviewed (approved):
Minor comment around handling of response from CFS when determining if the CFS configuration already exists.

jack-stanek-hpe commented:
I addressed the PR comments in [4542dd7](https://github.com/Cray-HPE/cfs-config-util/pull/7/commits/4542dd78e41ec315d2c3da0ad95fa23f19f51546), which also contains a small refactor to how `APIError`s are raised by the `APIClient` class so that callers can customize exception throwing. We might want to port that refactor back to SAT as part of [CRAYSAT-1539](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1539).

haasken-hpe commented on cfs_config_util/apiclient.py:
```diff
@@ -116,6 +116,31 @@ def __init__(self, session=None, host=None, cert_verify=None, timeout=None):
+        Raises:
+            APIError: containing an error message with details from the request
+        """
+        api_err_msg = (f"{req_type} request to URL '{url}' failed with status "
```
`req_type` and `url` aren't defined here.

haasken-hpe commented on cfs_config_util/apiclient.py:
```diff
@@ -116,6 +116,31 @@ def __init__(self, session=None, host=None, cert_verify=None, timeout=None):
+            APIError: containing an error message with details from the request
+        """
+        api_err_msg = (f"{req_type} request to URL '{url}' failed with status "
+                       f"code {r.status_code}: {r.reason}")
```
`r` is now `response`

haasken-hpe commented on cfs_config_util/cfs.py:
```diff
@@ -414,15 +414,19 @@ def save_to_cfs(self, name=None, overwrite=True):
+                elif response.status_code != 404:
+                    # If there's a failure for any reason other than a missing
+                    # configuration, throw an APIError
+                    self._cfs_client.raise_from_response(response)
```
I like the idea behind this change which appears to be intended to provide consistency with how the non-404 errors are represented, but I think we need to do something a little different given that `raise_from_response` doesn't have the `req_type` and `url` variables defined in that context. We could probably access the `response` object's `request.url` and `request.method` to get at those, but testing is needed to be sure that gets us what we want.

https://requests.readthedocs.io/en/latest/api/#requests.Response.request

jack-stanek-hpe replied:
> I like the idea behind this change which appears to be intended to provide consistency with how the non-404 errors are represented, but I think we need to do something a little different given that `raise_from_response` doesn't have the `req_type` and `url` variables defined in that context. We could probably access the `response` object's `request.url` and `request.method` to get at those, but testing is needed to be sure that gets us what we want.

I tested this this morning and it seems to get us what we want. (There aren't any unit tests for the API client in this repo yet, though I can add them if you'd like.) I've pushed a new fixup commit, fbec7d4, to address this.

haasken-hpe reviewed (changes requested):
I like the idea of the slight refactor, but there are some bugs to address.
